File: day16/day16.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum
{
	OP_ADDR,
	OP_ADDI,
	OP_MULR,
	OP_MULI,
	OP_BANR,
	OP_BANI,
	OP_BORR,
	OP_BORI,
	OP_SETR,
	OP_SETI,
	OP_GTIR,
	OP_GTRI,
	OP_GTRR,
	OP_EQIR,
	OP_EQRI,
	OP_EQRR,
	OP_COUNT,
};

static void compute(int *reg, int opcode, int a, int b, int c)
{
	switch (opcode)
	{
	case OP_ADDR: reg[c] = reg[a] + reg[b]; break;
	case OP_ADDI: reg[c] = reg[a] + b; break;
	case OP_MULR: reg[c] = reg[a] * reg[b]; break;
	case OP_MULI: reg[c] = reg[a] * b; break;
	case OP_BANR: reg[c] = reg[a] & reg[b];	break;
	case OP_BANI: reg[c] = reg[a] & b; break;
	case OP_BORR: reg[c] = reg[a] | reg[b]; break;
	case OP_BORI: reg[c] = reg[a] | b; break;
	case OP_SETR: reg[c] = reg[a]; break;
	case OP_SETI: reg[c] = a; break;
	case OP_GTIR: reg[c] = a > reg[b]; break;
	case OP_GTRI: reg[c] = reg[a] > b; break;
	case OP_GTRR: reg[c] = reg[a] > reg[b]; break;
	case OP_EQIR: reg[c] = a == reg[b]; break;
	case OP_EQRI: reg[c] = reg[a] == b; break;
	case OP_EQRR: reg[c] = reg[a] == reg[b]; break;
	}
}

static int process_sample(char m[OP_COUNT][OP_COUNT], int *before, int *instr, int *after)
{
	int count = 0;
	for (int i = 0; i < OP_COUNT; i++)
	{
		int reg[4];
		memmove(reg, before, sizeof(reg));
		compute(reg, i, instr[1], instr[2], instr[3]);
		if (memcmp(reg, after, sizeof(reg)) == 0)
		{
			count++;
			m[i][instr[0]] = 1;
		}
	}
	return count;
}

static int is_col_solved(const char m[OP_COUNT][OP_COUNT], int opcode)
{
	int count = 0, last = -1;
	for (int i = 0; i < OP_COUNT; i++)
	{
		if (m[i][opcode])
		{
			last = i;
		}
		count += m[i][opcode];
	}
	return count == 1 ? last : -1;
}

static void solve(char m[OP_COUNT][OP_COUNT], int opmap[OP_COUNT])
{
	memset(opmap, 0xff, sizeof(int) * OP_COUNT);
	for (int k = 0; k < OP_COUNT; k++)
	{
		int j = -1;
		for (int i = 0; i < OP_COUNT; i++)
		{
			if (opmap[i] == -1 && (j = is_col_solved(m, i)) >= 0)
			{
				opmap[i] = j;
				break;
			}
		}
		assert(j != -1);
		memset(m[j], 0, sizeof(m[j]));
	}
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s <filename>\n", argv[0]);
		return 1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "Cannot open %s\n", argv[1]);
		return 1;
	}

	char m[OP_COUNT][OP_COUNT];
	memset(m, 0, sizeof(m));

	size_t three_or_more = 0;
	for (;;)
	{
		int before[4];
		int instr[4];
		int after[4];
		int r = fscanf(input, " Before: [%d, %d, %d, %d]",
			       before+0, before+1, before+2, before+3);
		if (r != 4)
		{
			break;
		}
		r = fscanf(input, " %d %d %d %d",
			   instr+0, instr+1, instr+2, instr+3);
		if (r != 4)
		{
			break;
		}
		r = fscanf(input, " After: [%d, %d, %d, %d]",
			   after+0, after+1, after+2, after+3);
		if (r != 4)
		{
			break;
		}

		if (process_sample(m, before, instr, after) >= 3)
		{
			three_or_more++;
		}

	}
	printf("Part1: %zu\n", three_or_more);

	int opmap[OP_COUNT];
       	solve(m, opmap);

	/* execute the program */
	int op, a, b, c;
	int regs[4] = {0};
	while (fscanf(input, " %d %d %d %d", &op, &a, &b, &c) == 4)
	{
		compute(regs, opmap[op], a, b, c);
	}
	fclose(input);

	printf("Part2: %d\n", regs[0]);
	return 0;
}

File: day16/day16.py
#!/usr/bin/env python3

import sys

ADDR = 0
ADDI = 1
MULR = 2
MULI = 3
BANR = 4
BANI = 5
BORR = 6
BORI = 7
SETR = 8
SETI = 9
GTIR = 10
GTRI = 11
GTRR = 12
EQIR = 13
EQRI = 14
EQRR = 15
OPCOUNT = 16

def execute(op, a, b, c, regs):
    if op == ADDR: regs[c] = regs[a] + regs[b]
    elif op == ADDI: regs[c] = regs[a] + b
    elif op == MULR: regs[c] = regs[a] * regs[b]
    elif op == MULI: regs[c] = regs[a] * b
    elif op == BANR: regs[c] = regs[a] & regs[b]
    elif op == BANI: regs[c] = regs[a] & b
    elif op == BORR: regs[c] = regs[a] | regs[b]
    elif op == BORI: regs[c] = regs[a] | b
    elif op == SETR: regs[c] = regs[a]
    elif op == SETI: regs[c] = a
    elif op == GTIR: regs[c] = (a > regs[b]) and 1 or 0
    elif op == GTRI: regs[c] = (regs[a] > b) and 1 or 0
    elif op == GTRR: regs[c] = (regs[a] > regs[b]) and 1 or 0
    elif op == EQIR: regs[c] = (a == regs[b]) and 1 or 0
    elif op == EQRI: regs[c] = (regs[a] == b) and 1 or 0
    elif op == EQRR: regs[c] = (regs[a] == regs[b]) and 1 or 0
    else:
        assert 0, "Unknown operand {}".format(op)

def parse(txt):
    samples = []
    program = []
    before, after, instr = None, None, None
    emptylines = 0
    state = 0
    for line in txt.splitlines():
        if state == 0:
            if not line:
                if emptylines > 1:
                    state = 1
                else:
                    emptylines += 1
                continue
            elif line.startswith("Before:"):
                before = tuple(map(int, line[9:-1].split(",")))
            elif line.startswith("After:"):
                after = tuple(map(int, line[9:-1].split(",")))
                samples.append((before, instr, after))
            else:
                instr = tuple(map(int, line.split(" ")))
            emptylines = 0
        else:
            if not line:
                continue
            else:
                program.append(tuple(map(int, line.split(" "))))

    return samples, program

def process_samples(samples):
    m = [[0 for x in range(OPCOUNT)] for y in range(OPCOUNT)]
    three_or_more = 0
    for before, instr, after in samples:
        op, a, b, c = instr
        count = 0
        for i in range(16):
            regs = list(before)
            execute(i, a, b, c, regs)
            if tuple(regs) == after:
                count += 1
                m[i][op] = 1
        if count >= 3:
            three_or_more += 1

    return three_or_more, m

def is_col_solved(m, col):
    count, last = 0, -1
    for i, row in enumerate(m):
        if row[col]:
            count += 1
            last = i

    if count != 1:
        return -1

    return last

def solve(m):
    # constraint elimination
    opmap = [None] * OPCOUNT
    for _ in range(OPCOUNT):
        j = -1
        for i in range(OPCOUNT):
            if opmap[i] is None:
                j = is_col_solved(m, i)
                if j >= 0:
                    opmap[i] = j
                    break
        assert j != -1, "Cannot reduce the constraints"
        m[j] = [0] * OPCOUNT

    return opmap

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    samples, program = parse(txt)

    part1, m = process_samples(samples)
    print("Part1:", part1)

    opmap = solve(m)
    regs = [0] * 4
    for op, a, b, c in program:
        execute(opmap[op], a, b, c, regs)

    print("Part2:", regs[0])
